osu.Game/Modes/UI/HUDOverlay.cs

```diff
@@ -46,16 +39,17 @@ public virtual void OnHit(HitObject h)
 
         public virtual void OnMiss(HitObject h)
         {
-            ComboCounter?.Roll();
+            ComboCounter.Current.Value = 0;
             AccuracyCounter?.Set(AccuracyCounter.Count - 0.01f);
         }
 
-        protected ScoreOverlay()
+        protected HudOverlay(Ruleset ruleset)
         {
             RelativeSizeAxes = Axes.Both;
 
-            Children = new Drawable[] {
-                KeyCounter = CreateKeyCounter(),
+            Children = new Drawable[]
+            {
+                KeyCounter = CreateKeyCounter(ruleset.CreateGameplayKeys()),

```

Comment:
by adding this, you've now made HudOverlay dependent on ruleset (where it previously wasn't). We may want to just allow an outside caller to add keys directly to the KeyOverley to avoid this (means HudOverlay can be tested without creating a ruleset).